x-pack/plugins/fleet/public/applications/fleet/hooks/use_apm_service_href.ts

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { Overwrite } from 'utility-types';
+import useAsync from 'react-use/lib/useAsync';
+
+import { useLocator } from '../../../hooks/use_locator';
+import type { DataStream } from '../types';
+import { LOCATORS_IDS } from '../constants';
+
+type APMDataStream = Overwrite<
+  DataStream,
+  { package: 'apm'; serviceDetails: NonNullable<DataStream['serviceDetails']> }
+>;
+
+export const isAPMIntegration = (datastream: DataStream): datastream is APMDataStream =>
+  Boolean(datastream.package === 'apm' && datastream.serviceDetails);
+
+export const useAPMServiceHref = (datastream: DataStream) => {
+  const apmLocator = useLocator(LOCATORS_IDS.APM_LOCATOR);
+
+  const { error, loading, value } = useAsync(() => {
+    if (!isAPMIntegration(datastream)) return Promise.resolve();
+
+    const { serviceName, environment } = datastream.serviceDetails;
+
+    return apmLocator!.getUrl({
```

Comment:
I would prefer to throw an explicit error here, or not return a link if it's not set, rather than using the `!` operator. Both are fine with me. In what scenarios is the locator undefined?

Reply:
there are 2 ways of consuming a locator:
1. referencing the `ApmPluginSetup` directly inside the `fleet plugin setup` function as the locator is returned from the `APM plugin setup` function.
2. Getting the locator from the share plugin central registry dynamically through an ID (locators are registered in a map on runtime)

I went with option 2 to avoid importing APM code inside fleet plugin for circular dependency reasons,  the only way I see the locator being undefined is either when the APM plugin setup method fails or when misspelling the locator ID.